Redo developer install steps for happy path (#701)
medic/cht-docs: content/en/apps/guides/hosting/core-developer.md

---
title: "CHT Core Developer Hosting"
linkTitle: "CHT Core Developer Hosting"
weight: 40
description: >
  Compile the CHT Core from source when doing Core development
---

{{% alert title="Note" %}} This guide assumes you are a CHT Core developer wanting to compile the CHT Core from source code to make commits to the [public GitHub repository](https://github.com/medic/cht-core). To set up a hosting environment for devoloping apps, see the [app guide]({{< relref "apps/guides/hosting/app-developer.md" >}}).

To deploy the CHT in production, see either [AWS hosting]({{< relref "apps/guides/hosting/self-hosting.md" >}}) or [Self hosting]({{< relref "apps/guides/hosting/ec2-setup-guide.md" >}}){{% /alert %}}

## The Happy Path Installation

This CHT Core developer guide will have you install NodeJS, npm, Grunt and CouchDB (via Docker) on your local workstation. These instructions should work verbatim on Ubuntu 18-22, but will need tweaks for MacOS (via `brew`) or Windows (via WSL2).

### Install NodeJS, npm, grunt and Docker

First, update your current packages:

```shell

> sudo apt update&&sudo apt -y dist-upgrade

Reviewer: Is the "dist-" part necessary? I'd have just expected a `sudo apt upgrade`

Author: Based on my reading [here](https://linoxide.com/difference-between-apt-get-upgrade-and-dist-upgrade/) and [here](https://askubuntu.com/a/226213), it is now a best practice for me for all but production instances (of which I run none b/c personal servers are not "production" ;) . I think it fits here in dev land, so resolving, but happy to get some push back!

```

Then install `nvm`, add it to your path and install NodeJS 12:

```shell

> curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.34.0/install.sh | bash
> . ~/.bashrc

Reviewer: This will work for both zsh and bash.
```suggestion
curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.34.0/install.sh | $0
. ~/.$0rc
```

> nvm install v12

Reviewer: We can save money on that unnecessary 'v' character ;)

```suggestion
nvm install 12
```

```

Now let's ensure NodeJS 12 and npm 6 were installed. This should output version 12.x.x for NodeJS and 6.x.x for `npm`:

```shell

> node -v&&npm -v

Reviewer: Spaces for readability
 ```suggestion
node -v && npm -v
```

```

With NodeJS out of the way, let's install `grunt` via `npm` and supporting tools via `apt`:

```shell
npm install -g grunt-cli

> sudo apt -y install xsltproc curl uidmap jq

Author: Agreed! fixed in 6c9961e

```

> Install Docker with this sweet one liner (note it's directly from docker.com):

Reviewer: The other text is an unnecessary detail
```suggestion
Install Docker:
```

```shell
curl -fsSL get.docker.com -o get-docker.sh && sh get-docker.sh
```

It's easier if you don't always have to run `sudo` for all your Docker calls, so let's set that up:

```shell
dockerd-rootless-setuptool.sh install

> echo "export PATH=/usr/bin:$PATH" >> ~/.bashrc
> echo "export DOCKER_HOST=unix:///run/user/1000/docker.sock" >> ~/.bashrc
> . ~/.bashrc

Reviewer: Please consider us zsh people! :)
```suggestion
echo "export PATH=/usr/bin:$PATH" >> ~/.$0rc
echo "export DOCKER_HOST=unix:///run/user/1000/docker.sock" >> ~/.$0rc
. ~/.$0rc
```

Author: Just FYI:

```shell
➜  cht-docs git:(700-developer-install-happy-path) echo $0
zsh
```

(technically oh-my-z-sh)

```shell
docker run -d -p 5984:5984 -p 5986:5986 --name medic-couchdb -e COUCHDB_USER=medic -e COUCHDB_PASSWORD=password -v ~/cht-docker/local.d:/opt/couchdb/data -v ~/cht-docker/local.d:/opt/couchdb/etc/local.d apache/couchdb:2
```

Let's ensure CouchDB is set up with a test `curl` call. This should show "nonode@nohost" in JSON:

```shell
curl -X GET "http://medic:password@localhost:5984/_membership" | jq
```

Every time you run any `grunt` or `node` commands, it will expect `COUCH_NODE_NAME` and `COUCH_URL` environment variables to be set:

```shell
echo "export COUCH_NODE_NAME=nonode@nohost">> ~/.bashrc
echo "export COUCH_URL=http://medic:password@localhost:5984/medic">> ~/.bashrc
. ~/.bashrc
echo $COUCH_NODE_NAME&& echo $COUCH_URL
```

### CHT Core cloning and setup

Clone the main CHT Core repo from GitHub and initialize it with an `npm` command. Note this last command may take many minutes. Be patient!

```shell
git clone https://github.com/medic/cht-core ~/cht-core
cd ~/cht-core

> npm ci  

Reviewer: Please split out the `npm ci` call and instructions for it. It's different enough from the git clone. Then you can give a bit more detail on the `npm ci` with something like "install dependencies and perform other setup tasks".

Author: yup yup - good call! done in 3f4f766

```

We need to harden couch with a `grunt` call, required even in development:

```shell
grunt secure-couchdb
curl -X PUT "http://medic:password@localhost:5984/_node/$COUCH_NODE_NAME/_config/httpd/WWW-Authenticate" -d '"Basic realm=\"administrator\""' -H "Content-Type: application/json"
```

### Developing

> Now you have everything installed and can begin development! You'll need three separate terminals when doing development. In the first terminal, run:

Reviewer: >You'll need three separate terminals
I always appreciate things like this. It tells me what's about to happen and then when I'm doing it I know the context of what I'm doing. I know this is a really small case, but I still really like it :)

```shell
cd ~/cht-core&&grunt                            
```

Be **very** patient until you see:

> "Waiting..."

Then in a 2nd terminal run:

```shell

> cd ~/cht-core&&grunt dev-api

Reviewer: Spaces
```suggestion
cd ~/cht-core && grunt dev-api
```

```

Finally, in a 3rd terminal run:

```shell
cd ~/cht-core&&grunt dev-sentinel
```

That's it!  Now when you edit code in your IDE, it should automatically reload.  You can see the CHT running locally here: [http://localhost:5988/](http://localhost:5988/)

When you're done with development you can `ctrl + c` in the three terminals and stop the CouchDB container with `docker stop medic-couchdb`.  When you want to resume development later, run `docker start medic-couchdb` and re-run the three terminal commands.

## Other Paths

If you weren't able to follow the happy path above, here are some details about the developer install that may help you troubleshoot what went wrong.

### Prerequisites

If you had issues with following the above steps, check out these links for how to install the prerequisites on your specific platform:

* [Node.js 12.x](https://nodejs.org/)
* [npm 6.x.x](https://npmjs.com/)
* [grunt cli](https://gruntjs.com/using-the-cli)
* [xsltproc](http://www.sagehill.net/docbookxsl/InstallingAProcessor.html) 

> * [python 2.7](https://www.python.org/downloads/)

Author: Gotcha. I dug up [this ticket](https://github.com/medic/cht-core/issues/5608) which says maybe it won't work? I don't want to solve this now given how far reaching it might be.  I say we leave it as is and do nothing or leave it as is and file a "update requirements to be python 3" ticket.

Thoughts?

> * [python 2.7](https://www.python.org/downloads/)

Author: Meh - doing [the latter](https://github.com/medic/cht-core/issues/7636).

You can use `docker stop medic-couchdb` to stop it and `docker start medic-couchdb` to start it again. Remember that you'll need to start it whenever you restart your OS, which might not be the case if you use a normal OS package. `docker rm medic-couchdb` will totally remove the container.

Medic recommends you familiarise yourself with other Docker commands to make docker image and container management clearer.

### Required environment variables

Medic needs the following environment variables to be declared:
- `COUCH_URL`: the full authenticated url to the `medic` DB. Locally this would be  `http://myadminuser:myadminpass@localhost:5984/medic`
- `COUCH_NODE_NAME`: the name of your CouchDB's node. The Docker image default is `nonode@nohost`. Other installations may use `couchdb@127.0.0.1`. You can find out by querying [CouchDB's membership API](https://docs.couchdb.org/en/stable/api/server/common.html#membership)
- (optionally) `API_PORT`: the port API will run on. If not defined we use `5988`

> - (optionally) `CHROME_BIN`: only required if `grunt unit` or `grunt e2e` complain that they can't find Chrome.

Author: fixed in 8752c44

How to permanently define environment variables depends on your OS and shell (e.g. for bash you can put them `~/.bashrc`). You can temporarily define them with `export`:

> ```sh

Reviewer: ```suggestion
```shell
```

export COUCH_NODE_NAME=nonode@nohost
export COUCH_URL=http://myadminuser:myadminpass@localhost:5984/medic
```

## Tests

Refer to [the testing doc](https://github.com/medic/cht-core/blob/master/TESTING.md) in the GitHub repo.

## nginx-local-ip

> [`nginx-local-ip`](https://github.com/medic/nginx-local-ip) is a local proxy that keeps all traffic local, and runs without latency or throttling. If sharing your local CHT instance is not required, it is the preferred method to add a valid SSL certificate (rather than `ngrok` or `pagekite`).

Reviewer: Small wording changes
```suggestion
[`nginx-local-ip`](https://github.com/medic/nginx-local-ip) is a local proxy that keeps all traffic local, and runs without latency or throttling. If sharing your local CHT instance is not required, it is the recommended method to add a valid SSL certificate (rather than `ngrok` or similar).
```

```

## Remote Proxies

`ngrok` and `pagekite` are remote proxies that route local traffic between your client and the CHT via a remote SSL terminator. While easy and handy, they introduce latency and are sometimes throttled.

### ngrok

1. Create an [ngrok account](https://ngrok.com/), download and install the binary, then link your computer to your ngrok account.
1. Start `ngrok` to connect to:
  * The CHT API running via `grunt` or `horti`, execute `./ngrok http 5988`
  * The CHT API running via `docker`, execute `./ngrok http 443`
1. Access the app using the https address shown (e.g. `https://YOUR-NGROK-NAME.ngrok.io`, replacing `YOUR-NGROK-NAME` with what you signed up with).

**Note:** The service worker cache preload sometimes fails due to connection throttling (thereby causing an `ngrok` failure at startup).

### pagekite

1. Create a [pagekite account](https://pagekite.net/signup/), download and install the python script.
1. Start pagekite (be sure to replace `YOUR-PAGEKIT-NAME` with the URL you signed up for) to connect to:
  * The CHT API running via `grunt` or `horti`, execute `python2 pagekite.py 5988 YOUR-PAGEKIT-NAME.pagekite.me`

>   * The CHT API running via `docker`, execute `python2 pagekite.py 443 YOUR-PAGEKIT-NAME.pagekite.me`

Author: [nope](https://pagekite.net/downloads#s3linux)! fixed in c003704

linkTitle: "CHT Core Developer Setup"
weight: 40
description: >
  Getting your local machine ready to to development work on CHT Core.
---

{{% alert title="Note" %}} This guide assumes you are a CHT Core developer wanting to run the CHT Core from source code to make commits to the [public GitHub repository](https://github.com/medic/cht-core). To set up a your environment for developing apps, see the [app guide]({{< relref "apps/guides/hosting/app-developer.md" >}}).

To deploy the CHT in production, see either [AWS hosting]({{< relref "apps/guides/hosting/self-hosting.md" >}}) or [Self hosting]({{< relref "apps/guides/hosting/ec2-setup-guide.md" >}}){{% /alert %}}

## The Happy Path Installation

This CHT Core developer guide will have you install NodeJS, npm, Grunt and CouchDB (via Docker) on your local workstation. These instructions should work verbatim on Ubuntu 18-22, but will need tweaks for MacOS (via `brew`) or Windows (via WSL2).

### Install NodeJS, npm, grunt and Docker

First, update your current Ubuntu packages and install some supporting tools via `apt`:

```shell
sudo apt update && sudo apt -y dist-upgrade
sudo apt -y install xsltproc curl uidmap jq
```

> Then install `nvm`, add it to your path and install NodeJS 12:

Reviewer 2: How much do we want to explain _why_ we're doing this? Like, what even is nvm?

```shell

> curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.34.0/install.sh | $0

Reviewer 2: `0.34.0` is not the latest and the docs now recommend using `0.39.1`. But I wonder if we should just say `master` so we don't have to update this every time? Feels wrong though.

Author: Hmm - `curl` + `jq` = ftw? see 882adb2

. ~/.$0rc
nvm install 12
```

Now let's ensure NodeJS 12 and npm 6 were installed. This should output version 12.x.x for NodeJS and 6.x.x for `npm`:

```shell
node -v && npm -v
```

With NodeJS out of the way, let's install `grunt` via `npm`:

```shell
npm install -g grunt-cli
```

Install Docker:

```shell
curl -fsSL get.docker.com -o get-docker.sh && sh get-docker.sh
```

It's easier if you don't always have to run `sudo` for all your Docker calls, so let's set that up:

```shell
dockerd-rootless-setuptool.sh install
echo "export PATH=/usr/bin:$PATH" >> ~/.$0rc
echo "export DOCKER_HOST=unix:///run/user/1000/docker.sock" >> ~/.$0rc
. ~/.$0rc
```

In order for Docker to boot correctly, restart entire machine, which will complete the "Install" Section:

```shell
sudo reboot
```

### CouchDB Setup in Docker

Before we get started, let's run the simple `hello-world` Docker container. This will ensure docker is working as expected and output "Hello from Docker!" as well as some other intro text:

```shell
docker run hello-world

> ```

Reviewer 2: I love these verification steps. Fail early!

Now that we know Docker is set up, let's start our CouchDB container. Note this will run in the background and store its data in `/home/YOUR-USER/cht-docker`. The login for your CHT instance will be `medic` and the password will be `password`:

```shell
docker run -d -p 5984:5984 -p 5986:5986 --name medic-couchdb -e COUCHDB_USER=medic -e COUCHDB_PASSWORD=password -v ~/cht-docker/local.d:/opt/couchdb/data -v ~/cht-docker/local.d:/opt/couchdb/etc/local.d apache/couchdb:2
```

Let's ensure CouchDB is set up with a test `curl` call. This should show "nonode@nohost" in JSON:

```shell
curl -X GET "http://medic:password@localhost:5984/_membership" | jq
```

Every time you run any `grunt` or `node` commands, it will expect `COUCH_NODE_NAME` and `COUCH_URL` environment variables to be set:

```shell
echo "export COUCH_NODE_NAME=nonode@nohost">> ~/.$0rc
echo "export COUCH_URL=http://medic:password@localhost:5984/medic">> ~/.$0rc
. ~/.$0rc

> echo $COUCH_NODE_NAME && echo $COUCH_URL

Reviewer 2: Maybe split this out as a separate block so we can say "to make sure it's working, run this:"

Author: yup yup! done in 39f1bbb

```shell
npm ci
```

We need to harden couch with a `grunt` call, required even in development:

```shell
grunt secure-couchdb
curl -X PUT "http://medic:password@localhost:5984/_node/$COUCH_NODE_NAME/_config/httpd/WWW-Authenticate" -d '"Basic realm=\"administrator\""' -H "Content-Type: application/json"
```

### Developing

Now you have everything installed and can begin development! You'll need three separate terminals when doing development. In the first terminal, run:

```shell
cd ~/cht-core && grunt
```

Be **very** patient until you see:

> "Waiting..."

Then in a 2nd terminal run:

```shell
cd ~/cht-core && grunt dev-api
```

Finally, in a 3rd terminal run:

```shell
cd ~/cht-core && grunt dev-sentinel
```

That's it!  Now when you edit code in your IDE, it will automatically reload.  You can see the CHT running locally here: [http://localhost:5988/](http://localhost:5988/)

When you're done with development you can `ctrl + c` in the three terminals and stop the CouchDB container with `docker stop medic-couchdb`.  When you want to resume development later, run `docker start medic-couchdb` and re-run the three terminal commands.

> ## Other Paths

Reviewer 2: Maybe call this "Troubleshooting"? If you've forgotten (like I had) that the first heading was "Happy Path" this heading makes no sense.

Author: renamed AND linked - let's help find the right path! see 5a7ae9a

---
title: "CHT Core Developer Setup"
linkTitle: "CHT Core Developer Setup"
weight: 40
description: >
  Getting your local machine ready to to development work on CHT Core.
---

{{% alert title="Note" %}} This guide assumes you are a CHT Core developer wanting to run the CHT Core from source code to make commits to the [public GitHub repository](https://github.com/medic/cht-core). To set up a your environment for developing apps, see the [app guide]({{< relref "apps/guides/hosting/app-developer.md" >}}).

To deploy the CHT in production, see either [AWS hosting]({{< relref "apps/guides/hosting/self-hosting.md" >}}) or [Self hosting]({{< relref "apps/guides/hosting/ec2-setup-guide.md" >}}){{% /alert %}}

## The Happy Path Installation

This CHT Core developer guide will have you install NodeJS, npm, Grunt and CouchDB (via Docker) on your local workstation. These instructions should work verbatim on Ubuntu 18-22, but will need tweaks for MacOS (via `brew`) or Windows (via WSL2).

### Install NodeJS, npm, grunt and Docker

First, update your current Ubuntu packages and install some supporting tools via `apt`:

```shell
sudo apt update && sudo apt -y dist-upgrade

> sudo apt -y install xsltproc curl uidmap jq

Author: ok - rad - I'll commit this and then do another run through to make sure it all gels together.

> sudo apt -y install xsltproc curl uidmap jq

Reviewer: > ok - rad - I'll commit this and then do another run through to make sure it all gels together.

Thanks. I'll look for your confirmation that it all actually works and then I'm good to approve.
